File: src/test/java/io/bloco/faker/components/ColorTest.java
package io.bloco.faker.components;

import org.junit.Before;
import org.junit.Test;

import io.bloco.faker.Faker;

import static io.bloco.faker.test_helpers.RegularExpressionMatcher.matchesPattern;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.Matchers.lessThan;
import static org.hamcrest.Matchers.lessThanOrEqualTo;
import static org.hamcrest.core.AllOf.allOf;
import static org.hamcrest.core.Is.is;
import static org.hamcrest.number.OrderingComparison.greaterThanOrEqualTo;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertThat;

public class ColorTest {

    private Faker faker;

    @Before
    public void setUp() {
        faker = new Faker();
    }

    @Test
    public void hexColor() {
        assertThat(faker.color.hexColor(), matchesPattern("^#[0-9a-f]{6}$"));
    }

    @Test
    public void color() {
        assertNotNull(faker.color.colorName());
    }

    @Test
    public void singleRgbColor() {
        assertSingleRgbColor(faker.color.singleRgbColor());
    }

    @Test
    public void rgbColor() {
        int[] color = faker.color.rgbColor();
        assertThat(color.length, is(equalTo(3)));
        assertSingleRgbColor(color[0]);
        assertSingleRgbColor(color[1]);
        assertSingleRgbColor(color[2]);
    }

    @Test
    public void singleHslColor() {
        assertSingleHslColor(faker.color.singleHslColor());
    }

    @Test
    public void alphaChannel() {
        assertAlphaChannel(faker.color.alphaChannel());
    }

    @Test
    public void hslColor() {
        double[] color = faker.color.hslColor();
        assertThat(color.length, is(equalTo(3)));
        assertSingleHslColor(color[0]);
        assertSingleHslColor(color[1]);
        assertSingleHslColor(color[2]);
    }

    @Test
    public void hslaColor() {
        double[] color = faker.color.hslaColor();
        assertThat(color.length, is(equalTo(4)));
        assertSingleHslColor(color[0]);
        assertSingleHslColor(color[1]);
        assertSingleHslColor(color[2]);
        assertAlphaChannel(color[3]);
    }

    private void assertSingleRgbColor(int singleRgbColor) {
        assertThat(singleRgbColor, allOf(greaterThanOrEqualTo(0), lessThan(Color.MAX_RGB)));
    }

    private void assertSingleHslColor(double singleHslColor) {
        assertThat(singleHslColor, allOf(greaterThanOrEqualTo(0d), lessThan(Color.MAX_HSL)));
    }

    private void assertAlphaChannel(double alpha) {
        assertThat(alpha, allOf(greaterThanOrEqualTo(0d), lessThanOrEqualTo(1d)));
    }
}

File: lib/src/test/kotlin/io/bloco/faker/components/ColorTest.kt
package io.bloco.faker.components

import io.bloco.faker.Faker
import junit.framework.TestCase.assertNotNull
import junit.framework.TestCase.assertTrue
import org.junit.Before
import org.junit.Test

class ColorTest {
    private val faker = Faker()

    @Test
    fun hexColor() {
        assertTrue(faker.color.hexColor().matches(Regex("^#[0-9a-f]{6}\$")))
    }

    @Test
    fun color() {
        assertNotNull(faker.color.colorName())
    }

    @Test
    fun singleRgbColor() {
        assertSingleRgbColor(faker.color.singleRgbColor())
    }

    @Test
    fun rgbColor() {
        val color = faker.color.rgbColor()
        assertTrue(color.size == 3)
        assertSingleRgbColor(color[0])
        assertSingleRgbColor(color[1])
        assertSingleRgbColor(color[2])
    }

    @Test
    fun singleHslColor() {
        assertSingleHslColor(faker.color.singleHslColor())
    }

    @Test
    fun alphaChannel() {
        assertAlphaChannel(faker.color.alphaChannel())
    }

    @Test
    fun hslColor() {
        val color = faker.color.hslColor()
        assertTrue(color.size == 3)
        assertSingleHslColor(color[0])
        assertSingleHslColor(color[1])
        assertSingleHslColor(color[2])
    }

    @Test
    fun hslaColor() {
        val color = faker.color.hslaColor()
        assertTrue(color.size == 4)
        assertSingleHslColor(color[0])
        assertSingleHslColor(color[1])
        assertSingleHslColor(color[2])
        assertAlphaChannel(color[3])
    }

    private fun assertSingleRgbColor(singleRgbColor: Int) {
        assertTrue(singleRgbColor in 0..Color.MAX_RGB)
    }

    private fun assertSingleHslColor(singleHslColor: Double) {
        assertTrue(singleHslColor in 0.0..Color.MAX_HSL)
    }

    private fun assertAlphaChannel(alpha: Double) {
        assertTrue(alpha in 0.0..1.0)
    }
}